File: projects/dft_lammps_research/platform/api/sdks/python/dft_lammps.py
"""
DFT+LAMMPS Python SDK

Official Python client for the DFT+LAMMPS API Platform.

Installation:
    pip install dft-lammps-client

Quick Start:
    >>> from dft_lammps import Client
    >>> client = Client(api_key="your-api-key")
    >>> project = client.projects.create(name="My Project")
    >>> calculation = client.calculations.submit(
    ...     project_id=project.id,
    ...     structure="Li2S.cif",
    ...     calculation_type="dft"
    ... )
"""

__version__ = "1.0.0"

import os
import json
import time
import hmac
import hashlib
from typing import Optional, Dict, List, Any, Iterator, Union
from dataclasses import dataclass
from datetime import datetime
from urllib.parse import urljoin, urlencode

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


# Default configuration
DEFAULT_BASE_URL = "https://api.dft-lammps.org"
DEFAULT_TIMEOUT = 30
DEFAULT_MAX_RETRIES = 3


class DFTLAMMPSError(Exception):
    """Base exception for SDK errors"""
    pass


class AuthenticationError(DFTLAMMPSError):
    """Authentication failed"""
    pass


class RateLimitError(DFTLAMMPSError):
    """Rate limit exceeded"""
    pass


class NotFoundError(DFTLAMMPSError):
    """Resource not found"""
    pass


class ValidationError(DFTLAMMPSError):
    """Validation failed"""
    pass


@dataclass
class Project:
    """Project resource"""
    id: str
    name: str
    description: Optional[str]
    project_type: str
    status: str
    target_properties: Dict[str, Any]
    material_system: Optional[str]
    tags: List[str]
    total_structures: int
    completed_calculations: int
    failed_calculations: int
    created_at: datetime
    updated_at: Optional[datetime]
    owner_id: str


@dataclass
class Calculation:
    """Calculation resource"""
    id: str
    project_id: str
    calculation_type: str
    status: str
    structure: Dict[str, Any]
    parameters: Dict[str, Any]
    priority: int
    results: Optional[Dict[str, Any]]
    error_message: Optional[str]
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]


@dataclass
class Structure:
    """Structure resource"""
    id: str
    name: str
    format: str
    cell: List[List[float]]
    species: List[str]
    positions: List[List[float]]
    metadata: Dict[str, Any]
    created_at: datetime


@dataclass
class WebhookSubscription:
    """Webhook subscription"""
    webhook_id: str
    url: str
    events: List[str]
    status: str
    created_at: str


class HTTPClient:
    """HTTP client with retries and authentication"""
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = DEFAULT_BASE_URL,
        timeout: int = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES
    ):
        self.api_key = api_key or os.getenv("DFT_LAMMPS_API_KEY")
        if not self.api_key:
            raise AuthenticationError("API key required. Set DFT_LAMMPS_API_KEY env var or pass to constructor.")
        
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "User-Agent": f"dft-lammps-python/{__version__}",
        })
        
        # Configure retries
        retry_strategy = Retry(
            total=max_retries,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "OPTIONS", "POST", "PUT", "PATCH", "DELETE"]
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("https://", adapter)
        self.session.mount("http://", adapter)
    
    def _handle_response(self, response: requests.Response) -> dict:
        """Handle API response and errors"""
        if response.status_code == 200 or response.status_code == 201:
            return response.json() if response.content else {}
        elif response.status_code == 204:
            return {}
        elif response.status_code == 400:
            raise ValidationError(response.json().get("detail", "Validation failed"))
        elif response.status_code == 401:
            raise AuthenticationError("Invalid API key")
        elif response.status_code == 403:
            raise AuthenticationError("Permission denied")
        elif response.status_code == 404:
            raise NotFoundError("Resource not found")
        elif response.status_code == 429:
            raise RateLimitError("Rate limit exceeded. Try again later.")
        else:
            raise DFTLAMMPSError(f"API error: {response.status_code} - {response.text}")
    
    def request(
        self,
        method: str,
        endpoint: str,
        params: Optional[dict] = None,
        json_data: Optional[dict] = None,
        **kwargs
    ) -> dict:
        """Make HTTP request"""
        url = urljoin(self.base_url + "/", endpoint.lstrip("/"))
        
        response = self.session.request(
            method=method,
            url=url,
            params=params,
            json=json_data,
            timeout=self.timeout,
            **kwargs
        )
        
        return self._handle_response(response)
    
    def get(self, endpoint: str, params: Optional[dict] = None) -> dict:
        """GET request"""
        return self.request("GET", endpoint, params=params)
    
    def post(self, endpoint: str, json_data: Optional[dict] = None) -> dict:
        """POST request"""
        return self.request("POST", endpoint, json_data=json_data)
    
    def patch(self, endpoint: str, json_data: Optional[dict] = None) -> dict:
        """PATCH request"""
        return self.request("PATCH", endpoint, json_data=json_data)
    
    def delete(self, endpoint: str) -> dict:
        """DELETE request"""
        return self.request("DELETE", endpoint)


class ProjectsAPI:
    """Projects API client"""
    
    def __init__(self, client: HTTPClient):
        self.client = client
    
    def list(
        self,
        status: Optional[str] = None,
        project_type: Optional[str] = None,
        search: Optional[str] = None,
        page: int = 1,
        page_size: int = 20
    ) -> Dict[str, Any]:
        """List projects with pagination"""
        params = {"page": page, "page_size": page_size}
        if status:
            params["status"] = status
        if project_type:
            params["project_type"] = project_type
        if search:
            params["search"] = search
        
        return self.client.get("/api/v1/projects", params=params)
    
    def create(
        self,
        name: str,
        description: Optional[str] = None,
        project_type: str = "battery_screening",
        target_properties: Optional[Dict[str, Any]] = None,
        material_system: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> Project:
        """Create a new project"""
        data = {
            "name": name,
            "description": description,
            "project_type": project_type,
            "target_properties": target_properties or {},
            "material_system": material_system,
            "tags": tags or [],
        }
        
        response = self.client.post("/api/v1/projects", json_data=data)
        return Project(**response)
    
    def get(self, project_id: str) -> Project:
        """Get project by ID"""
        response = self.client.get(f"/api/v1/projects/{project_id}")
        return Project(**response)
    
    def update(self, project_id: str, **kwargs) -> Project:
        """Update project"""
        response = self.client.patch(f"/api/v1/projects/{project_id}", json_data=kwargs)
        return Project(**response)
    
    def delete(self, project_id: str) -> None:
        """Delete project"""
        self.client.delete(f"/api/v1/projects/{project_id}")


class CalculationsAPI:
    """Calculations API client"""
    
    def __init__(self, client: HTTPClient):
        self.client = client
    
    def submit(
        self,
        project_id: str,
        structure: Union[str, Dict[str, Any]],
        calculation_type: str = "dft",
        parameters: Optional[Dict[str, Any]] = None,
        priority: int = 5
    ) -> Calculation:
        """
        Submit a calculation.
        
        Args:
            project_id: Project ID
            structure: Structure file path or structure data dict
            calculation_type: Type of calculation (dft, lammps, ml)
            parameters: Calculation parameters
            priority: Priority (1-10)
        """
        # Handle file upload
        if isinstance(structure, str) and os.path.isfile(structure):
            with open(structure, 'r') as f:
                structure_data = f.read()
            # Upload structure first
            structure_obj = self._upload_structure(project_id, structure, structure_data)
            structure_dict = {
                "id": structure_obj.id,
                "format": structure_obj.format,
            }
        else:
            structure_dict = structure if isinstance(structure, dict) else {"data": structure}
        
        data = {
            "structure": structure_dict,
            "calculation_type": calculation_type,
            "parameters": parameters or {},
            "priority": priority,
        }
        
        response = self.client.post(
            f"/api/v1/projects/{project_id}/calculations",
            json_data=data
        )
        return Calculation(**response)
    
    def _upload_structure(
        self,
        project_id: str,
        file_path: str,
        data: str
    ) -> Structure:
        """Upload structure file"""
        ext = os.path.splitext(file_path)[1].lower().lstrip(".")
        name = os.path.basename(file_path)
        
        payload = {
            "name": name,
            "format": ext if ext in ["poscar", "cif", "xyz", "json"] else "poscar",
            "data": data,
        }
        
        response = self.client.post(
            f"/api/v1/projects/{project_id}/structures",
            json_data=payload
        )
        return Structure(**response)
    
    def get(self, calculation_id: str) -> Calculation:
        """Get calculation by ID"""
        response = self.client.get(f"/api/v1/calculations/{calculation_id}")
        return Calculation(**response)
    
    def list(
        self,
        project_id: str,
        status: Optional[str] = None,
        page: int = 1,
        page_size: int = 20
    ) -> List[Calculation]:
        """List calculations for a project"""
        params = {"page": page, "page_size": page_size}
        if status:
            params["status"] = status
        
        response = self.client.get(
            f"/api/v1/projects/{project_id}/calculations",
            params=params
        )
        return [Calculation(**c) for c in response.get("items", [])]
    
    def wait(
        self,
        calculation_id: str,
        timeout: Optional[int] = None,
        poll_interval: int = 5
    ) -> Calculation:
        """
        Wait for calculation to complete.
        
        Args:
            calculation_id: Calculation ID
            timeout: Maximum wait time in seconds
            poll_interval: Seconds between status checks
        
        Returns:
            Completed calculation
        
        Raises:
            TimeoutError: If timeout is reached
        """
        start_time = time.time()
        
        while True:
            calc = self.get(calculation_id)
            
            if calc.status in ["completed", "failed"]:
                return calc
            
            if timeout and (time.time() - start_time) > timeout:
                raise TimeoutError(f"Calculation {calculation_id} did not complete within {timeout}s")
            
            time.sleep(poll_interval)
    
    def submit_batch(
        self,
        project_id: str,
        calculations: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Submit batch of calculations"""
        data = {
            "project_id": project_id,
            "calculations": calculations,
        }
        return self.client.post(
            f"/api/v1/projects/{project_id}/calculations/batch",
            json_data=data
        )


class WebhooksAPI:
    """Webhooks API client"""
    
    def __init__(self, client: HTTPClient):
        self.client = client
    
    def subscribe(
        self,
        url: str,
        events: List[str],
        metadata: Optional[Dict[str, Any]] = None
    ) -> WebhookSubscription:
        """
        Subscribe to webhook events.
        
        Args:
            url: Webhook endpoint URL
            events: List of event types or ["*"] for all events
            metadata: Optional metadata
        """
        data = {
            "url": url,
            "events": events,
            "metadata": metadata or {},
        }
        
        response = self.client.post("/api/v1/webhooks/subscribe", json_data=data)
        return WebhookSubscription(**response)
    
    def list(self) -> List[WebhookSubscription]:
        """List webhook subscriptions"""
        response = self.client.get("/api/v1/webhooks")
        return [WebhookSubscription(**s) for s in response]
    
    def delete(self, webhook_id: str) -> None:
        """Delete webhook subscription"""
        self.client.delete(f"/api/v1/webhooks/{webhook_id}")
    
    @staticmethod
    def verify_signature(payload: str, signature: str, secret: str) -> bool:
        """
        Verify webhook signature.
        
        Args:
            payload: Raw request body
            signature: Signature from X-Webhook-Signature header
            secret: Webhook secret from subscription
        
        Returns:
            True if signature is valid
        """
        expected = hmac.new(
            secret.encode(),
            payload.encode(),
            hashlib.sha256
        ).hexdigest()
        expected_sig = f"sha256={expected}"
        
        return hmac.compare_digest(expected_sig, signature)


class Client:
    """
    DFT+LAMMPS API Client
    
    Main entry point for the SDK.
    
    Args:
        api_key: API key for authentication
        base_url: API base URL
        timeout: Request timeout in seconds
        max_retries: Maximum number of retries
    
    Example:
        >>> client = Client(api_key="your-key")
        >>> project = client.projects.create(name="My Project")
        >>> print(project.id)
    """
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = DEFAULT_BASE_URL,
        timeout: int = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES
    ):
        self._client = HTTPClient(
            api_key=api_key,
            base_url=base_url,
            timeout=timeout,
            max_retries=max_retries
        )
        
        # API resource clients
        self.projects = ProjectsAPI(self._client)
        self.calculations = CalculationsAPI(self._client)
        self.webhooks = WebhooksAPI(self._client)
    
    def health(self) -> Dict[str, Any]:
        """Check API health"""
        return self._client.get("/health")
    
    def usage(self) -> Dict[str, Any]:
        """Get usage statistics"""
        return self._client.get("/api/v1/usage")


# Convenience exports
__all__ = [
    "Client",
    "Project",
    "Calculation",
    "Structure",
    "WebhookSubscription",
    "DFTLAMMPSError",
    "AuthenticationError",
    "RateLimitError",
    "NotFoundError",
    "ValidationError",
]

File: projects/dft_lammps_research/platform/api/sdks/go/dftlammps.go
// Package dftlammps provides a Go client for the DFT+LAMMPS API Platform.
//
// Installation:
//   go get github.com/dft-lammps/go-client
//
// Quick Start:
//   client := dftlammps.NewClient("your-api-key")
//   project, err := client.Projects.Create(ctx, &CreateProjectRequest{
//       Name: "My Project",
//   })
//
package dftlammps

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"time"
)

const (
	defaultBaseURL    = "https://api.dft-lammps.org"
	defaultTimeout    = 30 * time.Second
	defaultMaxRetries = 3
	version           = "1.0.0"
)

// Client is the main API client
type Client struct {
	HTTPClient *http.Client
	BaseURL    string
	APIKey     string

	Projects     *ProjectsService
	Calculations *CalculationsService
	Webhooks     *WebhooksService
}

// ClientOption allows customization of the client
type ClientOption func(*Client)

// WithHTTPClient sets a custom HTTP client
func WithHTTPClient(httpClient *http.Client) ClientOption {
	return func(c *Client) {
		c.HTTPClient = httpClient
	}
}

// WithBaseURL sets a custom base URL
func WithBaseURL(baseURL string) ClientOption {
	return func(c *Client) {
		c.BaseURL = baseURL
	}
}

// NewClient creates a new API client
func NewClient(apiKey string, opts ...ClientOption) (*Client, error) {
	if apiKey == "" {
		apiKey = os.Getenv("DFT_LAMMPS_API_KEY")
	}
	if apiKey == "" {
		return nil, fmt.Errorf("API key required: set DFT_LAMMPS_API_KEY or pass to constructor")
	}

	client := &Client{
		HTTPClient: &http.Client{Timeout: defaultTimeout},
		BaseURL:    defaultBaseURL,
		APIKey:     apiKey,
	}

	for _, opt := range opts {
		opt(client)
	}

	// Initialize services
	client.Projects = &ProjectsService{client: client}
	client.Calculations = &CalculationsService{client: client}
	client.Webhooks = &WebhooksService{client: client}

	return client, nil
}

// doRequest makes an HTTP request with retries
func (c *Client) doRequest(ctx context.Context, method, path string, body interface{}, params url.Values) (*http.Response, error) {
	var bodyReader io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		bodyReader = bytes.NewReader(jsonBody)
	}

	// Build URL
	u, err := url.Parse(c.BaseURL + path)
	if err != nil {
		return nil, err
	}
	if params != nil {
		u.RawQuery = params.Encode()
	}

	// Create request
	req, err := http.NewRequestWithContext(ctx, method, u.String(), bodyReader)
	if err != nil {
		return nil, err
	}

	// Set headers
	req.Header.Set("Authorization", "Bearer "+c.APIKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "dft-lammps-go/"+version)

	// Execute with retries
	var resp *http.Response
	for attempt := 0; attempt <= defaultMaxRetries; attempt++ {
		resp, err = c.HTTPClient.Do(req)
		if err == nil && resp.StatusCode < 500 && resp.StatusCode != 429 {
			break
		}

		if attempt < defaultMaxRetries {
			time.Sleep(time.Duration(attempt+1) * time.Second)
		}
	}

	return resp, err
}

// decodeResponse decodes JSON response
func decodeResponse(resp *http.Response, v interface{}) error {
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return &APIError{
			StatusCode: resp.StatusCode,
			Message:    string(body),
		}
	}

	if v != nil {
		return json.NewDecoder(resp.Body).Decode(v)
	}

	return nil
}

// APIError represents an API error
type APIError struct {
	StatusCode int
	Message    string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("API error %d: %s", e.StatusCode, e.Message)
}

// IsNotFound returns true for 404 errors
func (e *APIError) IsNotFound() bool {
	return e.StatusCode == 404
}

// IsRateLimit returns true for 429 errors
func (e *APIError) IsRateLimit() bool {
	return e.StatusCode == 429
}

// IsAuthError returns true for 401/403 errors
func (e *APIError) IsAuthError() bool {
	return e.StatusCode == 401 || e.StatusCode == 403
}

// ==================== Projects API ====================

// ProjectsService handles project operations
type ProjectsService struct {
	client *Client
}

// Project represents a project
type Project struct {
	ID                    string                 `json:"id"`
	Name                  string                 `json:"name"`
	Description           string                 `json:"description"`
	ProjectType           string                 `json:"project_type"`
	Status                string                 `json:"status"`
	TargetProperties      map[string]interface{} `json:"target_properties"`
	MaterialSystem        string                 `json:"material_system"`
	Tags                  []string               `json:"tags"`
	TotalStructures       int                    `json:"total_structures"`
	CompletedCalculations int                    `json:"completed_calculations"`
	FailedCalculations    int                    `json:"failed_calculations"`
	CreatedAt             time.Time              `json:"created_at"`
	UpdatedAt             *time.Time             `json:"updated_at"`
	OwnerID               string                 `json:"owner_id"`
}

// ProjectList represents a paginated list of projects
type ProjectList struct {
	Items    []Project `json:"items"`
	Total    int       `json:"total"`
	Page     int       `json:"page"`
	PageSize int       `json:"page_size"`
	HasNext  bool      `json:"has_next"`
	HasPrev  bool      `json:"has_prev"`
}

// ListProjectsRequest represents list parameters
type ListProjectsRequest struct {
	Status      string
	ProjectType string
	Search      string
	Page        int
	PageSize    int
}

// List retrieves projects with pagination
func (s *ProjectsService) List(ctx context.Context, req *ListProjectsRequest) (*ProjectList, error) {
	params := url.Values{}
	if req.Page > 0 {
		params.Set("page", strconv.Itoa(req.Page))
	}
	if req.PageSize > 0 {
		params.Set("page_size", strconv.Itoa(req.PageSize))
	}
	if req.Status != "" {
		params.Set("status", req.Status)
	}
	if req.ProjectType != "" {
		params.Set("project_type", req.ProjectType)
	}
	if req.Search != "" {
		params.Set("search", req.Search)
	}

	resp, err := s.client.doRequest(ctx, "GET", "/api/v1/projects", nil, params)
	if err != nil {
		return nil, err
	}

	var result ProjectList
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}

	return &result, nil
}

// CreateProjectRequest represents project creation parameters
type CreateProjectRequest struct {
	Name             string                 `json:"name"`
	Description      string                 `json:"description,omitempty"`
	ProjectType      string                 `json:"project_type,omitempty"`
	TargetProperties map[string]interface{} `json:"target_properties,omitempty"`
	MaterialSystem   string                 `json:"material_system,omitempty"`
	Tags             []string               `json:"tags,omitempty"`
}

// Create creates a new project
func (s *ProjectsService) Create(ctx context.Context, req *CreateProjectRequest) (*Project, error) {
	resp, err := s.client.doRequest(ctx, "POST", "/api/v1/projects", req, nil)
	if err != nil {
		return nil, err
	}

	var project Project
	if err := decodeResponse(resp, &project); err != nil {
		return nil, err
	}

	return &project, nil
}

// Get retrieves a project by ID
func (s *ProjectsService) Get(ctx context.Context, projectID string) (*Project, error) {
	resp, err := s.client.doRequest(ctx, "GET", "/api/v1/projects/"+projectID, nil, nil)
	if err != nil {
		return nil, err
	}

	var project Project
	if err := decodeResponse(resp, &project); err != nil {
		return nil, err
	}

	return &project, nil
}

// UpdateProjectRequest represents project update parameters
type UpdateProjectRequest struct {
	Name             string                 `json:"name,omitempty"`
	Description      string                 `json:"description,omitempty"`
	Status           string                 `json:"status,omitempty"`
	TargetProperties map[string]interface{} `json:"target_properties,omitempty"`
	Tags             []string               `json:"tags,omitempty"`
}

// Update updates a project
func (s *ProjectsService) Update(ctx context.Context, projectID string, req *UpdateProjectRequest) (*Project, error) {
	resp, err := s.client.doRequest(ctx, "PATCH", "/api/v1/projects/"+projectID, req, nil)
	if err != nil {
		return nil, err
	}

	var project Project
	if err := decodeResponse(resp, &project); err != nil {
		return nil, err
	}

	return &project, nil
}

// Delete deletes a project
func (s *ProjectsService) Delete(ctx context.Context, projectID string) error {
	resp, err := s.client.doRequest(ctx, "DELETE", "/api/v1/projects/"+projectID, nil, nil)
	if err != nil {
		return err
	}

	return decodeResponse(resp, nil)
}

// ==================== Calculations API ====================

// CalculationsService handles calculation operations
type CalculationsService struct {
	client *Client
}

// Calculation represents a calculation
type Calculation struct {
	ID               string                 `json:"id"`
	ProjectID        string                 `json:"project_id"`
	CalculationType  string                 `json:"calculation_type"`
	Status           string                 `json:"status"`
	Structure        map[string]interface{} `json:"structure"`
	Parameters       map[string]interface{} `json:"parameters"`
	Priority         int                    `json:"priority"`
	Results          map[string]interface{} `json:"results"`
	ErrorMessage     string                 `json:"error_message"`
	CreatedAt        time.Time              `json:"created_at"`
	StartedAt        *time.Time             `json:"started_at"`
	CompletedAt      *time.Time             `json:"completed_at"`
}

// SubmitCalculationRequest represents calculation submission
type SubmitCalculationRequest struct {
	Structure       map[string]interface{} `json:"structure"`
	CalculationType string                 `json:"calculation_type"`
	Parameters      map[string]interface{} `json:"parameters,omitempty"`
	Priority        int                    `json:"priority,omitempty"`
}

// Submit creates a new calculation
func (s *CalculationsService) Submit(ctx context.Context, projectID string, req *SubmitCalculationRequest) (*Calculation, error) {
	resp, err := s.client.doRequest(ctx, "POST", "/api/v1/projects/"+projectID+"/calculations", req, nil)
	if err != nil {
		return nil, err
	}

	var calc Calculation
	if err := decodeResponse(resp, &calc); err != nil {
		return nil, err
	}

	return &calc, nil
}

// Get retrieves a calculation by ID
func (s *CalculationsService) Get(ctx context.Context, calculationID string) (*Calculation, error) {
	resp, err := s.client.doRequest(ctx, "GET", "/api/v1/calculations/"+calculationID, nil, nil)
	if err != nil {
		return nil, err
	}

	var calc Calculation
	if err := decodeResponse(resp, &calc); err != nil {
		return nil, err
	}

	return &calc, nil
}

// Wait polls until calculation completes
func (s *CalculationsService) Wait(ctx context.Context, calculationID string, pollInterval time.Duration) (*Calculation, error) {
	if pollInterval == 0 {
		pollInterval = 5 * time.Second
	}

	ticker := time.NewTicker(pollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-ticker.C:
			calc, err := s.Get(ctx, calculationID)
			if err != nil {
				return nil, err
			}

			if calc.Status == "completed" || calc.Status == "failed" {
				return calc, nil
			}
		}
	}
}

// ==================== Webhooks API ====================

// WebhooksService handles webhook operations
type WebhooksService struct {
	client *Client
}

// WebhookSubscription represents a webhook subscription
type WebhookSubscription struct {
	WebhookID string            `json:"webhook_id"`
	URL       string            `json:"url"`
	Events    []string          `json:"events"`
	Secret    string            `json:"secret"`
	Status    string            `json:"status"`
	CreatedAt string            `json:"created_at"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

// Subscribe creates a new webhook subscription
func (s *WebhooksService) Subscribe(ctx context.Context, url string, events []string, metadata map[string]string) (*WebhookSubscription, error) {
	req := map[string]interface{}{
		"url":      url,
		"events":   events,
		"metadata": metadata,
	}

	resp, err := s.client.doRequest(ctx, "POST", "/api/v1/webhooks/subscribe", req, nil)
	if err != nil {
		return nil, err
	}

	var sub WebhookSubscription
	if err := decodeResponse(resp, &sub); err != nil {
		return nil, err
	}

	return &sub, nil
}

// List retrieves all webhook subscriptions
func (s *WebhooksService) List(ctx context.Context) ([]WebhookSubscription, error) {
	resp, err := s.client.doRequest(ctx, "GET", "/api/v1/webhooks", nil, nil)
	if err != nil {
		return nil, err
	}

	var subs []WebhookSubscription
	if err := decodeResponse(resp, &subs); err != nil {
		return nil, err
	}

	return subs, nil
}

// Delete removes a webhook subscription
func (s *WebhooksService) Delete(ctx context.Context, webhookID string) error {
	resp, err := s.client.doRequest(ctx, "DELETE", "/api/v1/webhooks/"+webhookID, nil, nil)
	if err != nil {
		return err
	}

	return decodeResponse(resp, nil)
}

// VerifySignature verifies a webhook signature
func VerifySignature(payload []byte, signature, secret string) bool {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(payload)
	expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	return hmac.Equal([]byte(expected), []byte(signature))
}

// ==================== Utility Functions ====================

// Health checks API health
func (c *Client) Health(ctx context.Context) (map[string]interface{}, error) {
	resp, err := c.doRequest(ctx, "GET", "/health", nil, nil)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}

	return result, nil
}

// Usage retrieves usage statistics
func (c *Client) Usage(ctx context.Context) (map[string]interface{}, error) {
	resp, err := c.doRequest(ctx, "GET", "/api/v1/usage", nil, nil)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}

	return result, nil
}
